dipy/tracking/fbcmeasures.pyx

```diff
@@ -0,0 +1,334 @@
+                        for pointId2 in range(streamlines_length[lineId2] - 1):
+                            # compute displacement
+                            xd_mp[lineId] = int(streamlines[lineId, pointId, 0] 
+                            - streamlines[lineId2, pointId2, 0] + 0.5)
+                            yd_mp[lineId] = int(streamlines[lineId, pointId, 1] 
+                            - streamlines[lineId2, pointId2, 1] + 0.5)
+                            zd_mp[lineId] = int(streamlines[lineId, pointId, 2] 
+                            - streamlines[lineId2, pointId2, 2] + 0.5)
+                            
+                            # if position is outside the kernel bounds, skip
+                            if xd_mp[lineId] > hn or -xd_mp[lineId] > hn or \
+                               yd_mp[lineId] > hn or -yd_mp[lineId] > hn or \
+                               zd_mp[lineId] > hn or -zd_mp[lineId] > hn:
+                                continue
+                            
+                            # grab kernel value from LUT
+                            score_mp[lineId] += \
+                                        lut[streamlines_nearestp[lineId, pointId], 
+                                        streamlines_nearestp[lineId2, pointId2], 
+                                        hn+xd_mp[lineId], 
+                                        hn+yd_mp[lineId], 
+                                        hn+zd_mp[lineId]]  # ang_v, ang_r, x, y, z
+                                    
+                    streamline_scores[lineId, pointId] = score_mp[lineId]
+
+        if have_openmp and num_threads is not None:
+            openmp.omp_set_num_threads(all_cores)
+        
+        # Save LFBC as class member
+        self.streamlines_lfbc = streamline_scores
+        
+        # compute RFBC for each fiber
+        self.streamlines_rfbc = compute_rfbc(streamlines_length, 
+                                              streamline_scores)
+        
+def compute_rfbc(streamlines_length, streamline_scores):
+    """ Compute the relative fiber to bundle coherence (RFBC)
+
+    Parameters
+    ----------
+    streamlines_length : 1D int array
+        Contains the length of each streamline
+    streamlines_scores : 2D double array
+        Contains the local fiber to bundle coherence (LFBC) for each streamline 
+        element.
+
+    Returns
+    ----------
+    output: normalized lowest average LFBC region along the fiber
+    """
+
+    # finds the region of the fiber with minimal length if 7 points in which the
+    # LFBC is the lowest
+    intLength = min(np.amin(streamlines_length), 7)
+    intValue = np.apply_along_axis(lambda x: min_moving_average(x[~np.isnan(x)], intLength), 
+                                    1, streamline_scores)
+    averageTotal = np.mean(np.apply_along_axis(
+                lambda x:np.mean(np.extract(x[~np.isnan(x)] >= 0, x[~np.isnan(x)])), 1, streamline_scores))
+    if not averageTotal == 0:
```

Comment:
Gotcha : I missed the effect of that call to np.mean! Makes perfect sense.

On Thu, Mar 31, 2016 at 6:45 AM, Stephan Meesters <notifications@github.com>
wrote:

> In dipy/tracking/fbcmeasures.pyx
> <https://github.com/nipy/dipy/pull/828#discussion_r58055178>:
>
> > +        Contains the local fiber to bundle coherence (LFBC) for each streamline
> > +        element.
> > +
> > +    Returns
> > +    ----------
> > +    output: normalized lowest average LFBC region along the fiber
> > +    """
> > +
> > +    # finds the region of the fiber with minimal length if 7 points in which the
> > +    # LFBC is the lowest
> > +    intLength = min(np.amin(streamlines_length), 7)
> > +    intValue = np.apply_along_axis(lambda x: min_moving_average(x[~np.isnan(x)], intLength),
> > +                                    1, streamline_scores)
> > +    averageTotal = np.mean(np.apply_along_axis(
> > +                lambda x:np.mean(np.extract(x[~np.isnan(x)] >= 0, x[~np.isnan(x)])), 1, streamline_scores))
> > +    if not averageTotal == 0:
>
> If I recall correctly, apply_along_axis is applied to streamline_scores
> which is a 2D array. It is applied to the first axis and by doing so loops
> over each fiber, and then applies the lambda function. This results in a 1D
> array of moving average values, and finally np.mean is used to get the
> total average of all moving average values (confusing I know..). In the end
> average_total is a number.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/828/files/c0cdf42759434257b496b818399ebbd38f0a9627#r58055178>
>